_posts/2019-02-15-introducing-aquery.md

```diff
@@ -0,0 +1,116 @@
+
+Enter `aquery`.
+
+![bazel queries and phases](/assets/bazel-queries.png)
+
+`aquery` runs on the Configured Target Graph and queries the Action Graph. The Action Graph<sup>[2](#footnote2)</sup> is the result of the Analysis Phase. It is a bipartite graph with the following types of nodes:
+
+- Artifacts: either a source file or any output file produced by an action
+- Actions: the functional step that takes a list of Artifacts as input and outputs a list of Artifacts. Note that any (output) artifact is produced by exactly one action. The Action Graph conveys explicit step-by-step instructions on how the build would be executed.
+
+With `aquery`, it is now possible to tap into that knowledge.
+
+## `bazel aquery`
+
+`aquery` is useful when we are interested in the properties of the Actions/Artifacts in the Action Graph. The basic structure of `aquery` output is as follows:
+
+```
+$ bazel aquery '//some:label'
+action 'Writing file some_file_name'
+  Mnemonic: ...
+  Owner: ...
+  Configuration: ...
+  ActionKey: ...
+  Inputs: [...]
+  Outputs: [...]
+...
+```
+
+Each action entry encapsulates all the information you need to know about how this action is to be executed: the actual commands run, the configuration in which the action is run, its input/output artifacts, and other attributes.
+
+Another nifty feature in `aquery` is the ability to filter the actions based on their inputs, outputs and mnemonics. This is useful to answer questions like: “Which action, from which target, is responsible for creating file foo.out”.
+
+```
+# List all actions generated while building all dependencies of //src/target_a
+$ bazel aquery 'deps(//src/target_a)'
+
+# List all actions generated while building all dependencies of //src/target_a
+# that have C++ files in their inputs.
+$ bazel aquery 'inputs(".*cc”, deps(//src/target_a))'
+
+# Which action generated `foo.out` after building all dependencies of target //src/target_a
+$ bazel aquery ‘outputs(“.*foo.out”, deps(//src/target_a))’
+```
+
+Apart from these basic features, `aquery` offers customizations for your specific use cases with its various flags and tools.
+
+### Output Formats
+`aquery` supports 3 different output formats: `text` (default, human-readable with formatting), `proto` and `textproto` (a human-readable representation of the proto output).
+
+### `--skyframe_state`
+A common use case of `aquery` is to find the action responsible for generating a particular file `foo.out`. However, it is often the case that multiple build commands for different targets were run prior to the query, and the user might not remember all the targets that were built.
+
+The flag `--skyframe_state` comes in handy in such situations.
+
+```
+# Find all actions on skyframe that has “foo.out” as an output
+bazel aquery --skyframe_state --output=proto ‘outputs(“*.foo.out”)’
```

Comment:
My reading of this section is that `bazel aquery --output=proto 'outputs("*.foo.out")'` would also be a valid query, somehow different from that writing with `--skyframe_state`? Can you clarify a bit more precisely exactly what this flag does?

Reply:
Our intention for this blog post was just to give a brief overview, while more details are explained in the documentation. This is a fair point though, and I'll add a sentence to highlight the difference in query syntax with and without `--skyframe_state`.